### PR TITLE
PETScWrappers::MPI::SparseMatrix: use SEQAIJ with one process

### PR DESCRIPTION
**Pull request by stefanozampini**
With one process, it is more efficient to use `MATSEQAIJ`

So, use `MATAIJ` to create the object (which is a proxy for `SEQAIJ` in sequential and `MPIAIJ` in parallel) and add missing preallocators

### CODE REVIEW

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 1]
```diff
@@ -300,12 +300,19 @@ namespace PETScWrappers
                                               rowstart_in_window.data(),
                                               colnums_in_window.data(),
                                               nullptr);
+          ierr = MatSeqAIJSetPreallocationCSR(matrix,
+                                              rowstart_in_window.data(),
+                                              colnums_in_window.data(),
+                                              nullptr);
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc:306]: This now calls first `MatMPIAIJSetPreallocationCSR` and then `MatSeqAIJSetPreallocationCSR`, with the same arguments. Is that what you intended? If so, would you mind updating the comment above these two calls to explain why calling both makes sense?

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 2]
```diff
@@ -300,12 +300,19 @@ namespace PETScWrappers
...
           AssertThrow(ierr == 0, ExcPETScError(ierr));
         }
       else
         {
           PetscInt i = 0;
-          ierr       = MatMPIAIJSetPreallocationCSR(matrix, &i, &i, nullptr);
+
+          ierr = MatSeqAIJSetPreallocationCSR(matrix, &i, &i, nullptr);
+          AssertThrow(ierr == 0, ExcPETScError(ierr));
+          ierr = MatMPIAIJSetPreallocationCSR(matrix, &i, &i, nullptr);
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc:315]: And then same here

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 3]
```diff
@@ -419,6 +426,10 @@ namespace PETScWrappers
           // then call the petsc function
           // that summarily allocates these
           // entries:
+          ierr = MatSeqAIJSetPreallocationCSR(matrix,
+                                              rowstart_in_window.data(),
+                                              colnums_in_window.data(),
+                                              nullptr);
           ierr = MatMPIAIJSetPreallocationCSR(matrix,
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc:433]: And here

File: source/lac/petsc_parallel_sparse_matrix.cc (hunk 1 above)
Comment (stefanozampini) [reply source/lac/petsc_parallel_sparse_matrix.cc:306]: Yes, this is how PETSc polymorphism works. An `MPIAIJ` matrix does not implement `SeqAIJSetPreallocationCSR` and vice-versa. So for example, if the matrix is of type `MPIAIJ` the `SeqAIJSet...` call is a dummy call. I will add a comment

**Comment by stefanozampini** [comment]
@bangerth This is ready to go. I have implemented the comments you requested

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 4]
```diff
@@ -418,7 +435,15 @@ namespace PETScWrappers
 
           // then call the petsc function
           // that summarily allocates these
-          // entries:
+          // entries.
+          // Here we both call the specific API since this is how
+          // PETSc polymorphism works. If the matrix is of type MPIAIJ,
+          // the second call is dummy. If the matrix is of type SEQAIJ,
+          // the first call is dummy.
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc:442]: Ah, fascinating. It's not a design I would have chosen, but ok :-)

**Review by bangerth — APPROVED** [review]

**Comment by tamiko** [comment]
We have an unrelated intermittend failure of `mpi/petsc_step_27.mpirun=2`. I observe this on a number of configurations at the moment.

**Pull request merged**